### PR TITLE
Ops 5077/procurement dashboard details

### PR DESCRIPTION
**Pull request by Santi-3rd**
## What changed

Adds the details section to the Procurement Dashboard. 

## Issue

#5077 

## How to test

- All tests pass
- Select different options in the Proc Shop filter and check if the tables within the accordions are filtered correctly.
- Select one of the agreements within the accordions and advance its procurement tracker to its next step. Check the procurement dashboard again to see if it reflects the agreement's new step.

## A11y impact

- [ ] No accessibility-impacting changes in this PR
- [ ] Accessibility changes included and validated against WCAG 2.1 AA intent
- [ ] Any temporary suppression includes `A11Y-SUPPRESSION` metadata (owner, expires, rationale)

## Screenshots

image
image

## Definition of Done Checklist
- [x] OESA: Code refactored for clarity
- [x] OESA: Dependency rules followed
- [x] Automated unit tests updated and passed
- [x] Automated integration tests updated and passed
- [x] Automated quality tests updated and passed
- [x] Automated load tests updated and passed
- [x] Automated a11y tests updated and passed
- [x] Automated security tests updated and passed
- [x] 90%+ Code coverage achieved
- [-]Form validations updated

### CODE REVIEW

**Review by johndeange — COMMENTED** [review]
## Code Review: Procurement Dashboard Details

### Overview

Adds a details section to the Procurement Dashboard, displaying agreements grouped by procurement step (Acquisition Planning through Award) in sortable, expandable accordions. Includes:
- Backend: new `_compute_days_in_procurement_step` service function + API surface
- Frontend: new `details/` component tree (accordion, step summary, table, expandable row)
- Data: additional test agreements/trackers/BLIs for the dashboard
- File reorganization: existing summary components moved to a `summary/` subdirectory

**CI: All checks passing.**

---

### Issues

**1. Unused `fiscal_year` parameter in `_compute_days_in_procurement_step`**

The function signature accepts `fiscal_year: int | None` but never uses it. The existing `_compute_procurement_step_summary` does the same, so this may be an intentional placeholder for future filtering, but it's worth clarifying.

(`backend/ops_api/ops/services/agreements.py:1048`)

**2. `useSortData` parameter overloading**

The 5th parameter of `useSortData` is named `totalFunding` but `ProcurementDetailsTable` passes an object (`sortContext`). This works because dispatch is type-based, but it's confusing for readers:

```javascript
// use-sortable-data.hooks.js:120
export const useSortData = (items, descending, sortCondition, sortType, totalFunding = 0) => {
```

Consider renaming to `context` or a more generic name now that it serves dual purposes.

**3. Typo in STEP_DESCRIPTIONS (Award)**

`frontend/src/pages/procurementDashboard/details/ProcurementDetailsStep.jsx:28`:
```
"...before changing the agreement to Awarded in OPS (this will also change the budget lines in executing status to obligated status).."
```
Double period at the end.

**4. `proc_shop_fee_percentage` inconsistency in test data**

Some new BLIs use `0.005` and `0.048` as fee percentages. Per project convention, fee percentages should be whole numbers (e.g., `5.0` = 5%). Values like `0.005` would be interpreted as 0.005% (essentially zero). If these are meant to be 0.5% and 4.8%, they should be `0.5` and `4.8`. If intentionally near-zero for test purposes, that's fine — just noting the inconsistency.

---

### Suggestions (non-blocking)

**5. `DetailsBuilderAccordion` duplicates `StepBuilderAccordion`**

The new component imports the same CSS and replicates the same heading structure as `StepBuilderAccordion` with fewer props. Consider extending the existing component with an optional variant prop instead of a near-duplicate.

**6. Hardcoded "TBD" fields in expanded row**

`ProcurementDetailsTableRow.jsx` renders "TBD" and `$0.00` for Initial/Final Req fields. If these are known future data points, a comment or TODO would help. If they'll never have data, consider removing them to reduce noise.

**7. `target_completion_date` derivation**

In `ProcurementDetails.jsx:53`, the target date is pulled from `step.target_completion_date` on each tracker step. If this field doesn't exist on all step models, it would silently be `null`. The backend `_compute_days_in_procurement_step` uses `step_completed_date` and `step_start_date` instead — confirm the frontend field name matches what the API actually returns.

---

### Positives

- Clean component decomposition with clear data flow
- Comprehensive unit test coverage for the new backend function (edge cases for inactive trackers, out-of-range steps, missing start dates)
- Good use of `useMemo` for derived data in the frontend
- The file reorganization into `summary/` and `details/` is a good structural improvement
- The polymorphic subclass resolution fix in `import_data.py` is a solid improvement

---

### Summary

The main functional concerns are the fee percentage values (#4) and the `target_completion_date` field name (#7) — worth verifying those are intentional. The rest are minor quality items. Overall this is a well-structured PR with good test coverage.

**Review by josbell — CHANGES_REQUESTED** [review]
## Review Summary

Thanks for the comprehensive work on the procurement dashboard details! I've identified **3 bugs** that need to be fixed before merging. All are straightforward fixes that should take ~10 minutes total.

### ✅ Overall Assessment
- Architecture and component structure look solid
- Good test coverage for most functionality
- Clean separation of concerns

### ⚠️ Issues Requiring Fixes

#### 1. **CRITICAL: Fee Percentage Format Violation**
**Files**: `backend/data_tools/data/agreements_and_blin_data.json5`
**Lines**: 1067, 1083, 1099, 1115, 1130, 1145, 1159, 1173, 1188

**Issue**: Test data uses decimal format (0.005, 0.048) instead of whole numbers as required by the project convention.

**Why it's wrong**: The frontend `calculateTotal` helper divides by 100 internally, so:
- `0.005` becomes `amount * 0.005 / 100` = **0.005% fee** (not 0.5%)
- `0.048` becomes `amount * 0.048 / 100` = **0.048% fee** (not 4.8%)

**Evidence**: 
- CLAUDE.md states: "Fee percentages are stored and passed as whole numbers (e.g., `5.0` = 5%, not `0.05`)"
- Your own newer entries (lines 17986, 18001, 18016) correctly use `5.0`, `4.8`, `4.5`

**Fix**: Change all decimal values to whole numbers:
```diff
- "proc_shop_fee_percentage": 0.005,
+ "proc_shop_fee_percentage": 0.5,

- "proc_shop_fee_percentage": 0.048,
+ "proc_shop_fee_percentage": 4.8,
```

---

#### 2. **HIGH: Null Reference Will Crash**
**File**: `frontend/src/pages/procurementDashboard/details/ProcurementDetailsTableRow.jsx`
**Line**: 71

**Issue**: Accessing `agreement.procurement_shop.abbr` without optional chaining will crash when `procurement_shop` is null.

**Evidence**:
- Backend model allows null: `awarding_entity_id: Mapped[Optional[int]]`
- Your own test includes: `procurement_shop: null` (ProcurementDashboardPage.test.jsx line 198)
- All other components use optional chaining: `AgreementBLIAccordion.jsx`, `AgreementDetailsView.jsx`, `ProcurementDashboardPage.jsx`

**Fix**: Add optional chaining:
```diff
- <td data-cy="proc-shop">{agreement.procurement_shop.abbr || ""}</td>
+ <td data-cy="proc-shop">{agreement.procurement_shop?.abbr || ""}</td>
```

---

#### 3. **MEDIUM: Invalid Date Handling in Sorting**
**File**: `frontend/src/hooks/use-sortable-data.hooks.js`
**Lines**: 48-49

**Issue**: Creating Date objects without validation causes unpredictable sorting:
- `new Date(null)` → 1970-01-01 (sorts as earliest, misleading)
- `new Date(undefined)` → Invalid Date (NaN comparisons fail)

**Fix**: Validate before creating Date:
```diff
  case tableSortCodes.budgetLineCodes.OBLIGATE_BY:
-     return new Date(budgetLine.date_needed);
+     const dateStr = budgetLine.date_needed;
+     if (!dateStr) return new Date(0);
+     const parsed = new Date(dateStr);
+     return isNaN(parsed.getTime()) ? new Date(0) : parsed;
```

---

### ✨ What's Good
- Excellent test coverage for the new components
- Clean component decomposition
- Proper use of RTK Query patterns
- Backend eager loading is correctly configured (no N+1 queries)

Once these 3 fixes are in, this is ready to merge!

File: backend/data_tools/data/agreements_and_blin_data.json5 [hunk 1]
```diff
@@ -694,6 +694,98 @@
       psc_contract_specialist: "PSC Contract Specialist",
```
Comment (josbell) [inline backend/data_tools/data/agreements_and_blin_data.json5:694]: **CRITICAL: Fee percentage format violation**

Change `0.005` to `0.5` (represents 0.5%, not 0.005%).

The frontend `calculateTotal` helper divides by 100 internally, so this will calculate as 0.005% instead of 0.5%.

Also fix the same issue on lines 1083, 1099, 1115, 1130, 1145, 1159, 1173, 1188.

File: frontend/src/pages/procurementDashboard/details/ProcurementDetailsTableRow.jsx [hunk 2]
```diff
@@ -0,0 +1,305 @@
+import { useMemo } from "react";
```
Comment (josbell) [inline frontend/src/pages/procurementDashboard/details/ProcurementDetailsTableRow.jsx:1]: **HIGH: Null reference will crash**

Add optional chaining: `agreement.procurement_shop?.abbr || ""`

Without the `?`, this will throw "Cannot read property 'abbr' of null" when an agreement has no procurement shop assigned.

Evidence: Backend allows null (`awarding_entity_id: Mapped[Optional[int]]`), and all other components use optional chaining (`AgreementBLIAccordion.jsx`, `AgreementDetailsView.jsx`, etc.).

File: frontend/src/hooks/use-sortable-data.hooks.js [hunk 3]
```diff
@@ -11,7 +11,8 @@ export const SORT_TYPES = {
     BUDGET_LINES: "Budget Lines",
```
Comment (josbell) [inline frontend/src/hooks/use-sortable-data.hooks.js:11]: **MEDIUM: Invalid date handling**

`new Date(null)` returns 1970-01-01 (sorts as earliest), and `new Date(undefined)` returns Invalid Date (unpredictable sorting).

Add validation:
```javascript
const dateStr = budgetLine.date_needed;
if (!dateStr) return new Date(0);
const parsed = new Date(dateStr);
return isNaN(parsed.getTime()) ? new Date(0) : parsed;
```

**Review by josbell — CHANGES_REQUESTED** [review]
## Review Summary

Thanks for the comprehensive work on the procurement dashboard details! I've identified **3 bugs** that need to be fixed before merging. All are straightforward fixes that should take ~10 minutes total.

### ✅ Overall Assessment
- Architecture and component structure look solid
- Good test coverage for most functionality
- Clean separation of concerns

### ⚠️ Issues Requiring Fixes

See inline comments for specific line-level fixes needed.

#### 1. **CRITICAL: Fee Percentage Format Violation**
**Lines**: 1067, 1083, 1099, 1115, 1130, 1145, 1159, 1173, 1188 in `backend/data_tools/data/agreements_and_blin_data.json5`

Test data uses decimal format (0.005, 0.048) instead of whole numbers. The frontend `calculateTotal` helper divides by 100 internally, so `0.005` becomes **0.005% fee** (not 0.5%).

**Fix**: Change `0.005` → `0.5`, `0.048` → `4.8`

#### 2. **HIGH: Null Reference Will Crash**
**Line 71** in `frontend/src/pages/procurementDashboard/details/ProcurementDetailsTableRow.jsx`

Accessing `agreement.procurement_shop.abbr` without optional chaining will crash when `procurement_shop` is null.

**Fix**: Add optional chaining: `agreement.procurement_shop?.abbr`

#### 3. **MEDIUM: Invalid Date Handling**
**Lines 48-49** in `frontend/src/hooks/use-sortable-data.hooks.js`

`new Date(null)` → 1970-01-01 (misleading), `new Date(undefined)` → Invalid Date (unpredictable sorting).

**Fix**: Validate dates before creating Date objects.

---

### ✨ What's Good
- Excellent test coverage
- Clean component decomposition
- Proper RTK Query patterns
- Backend eager loading correctly configured

Once these 3 fixes are in, this is ready to merge!

**Review by josbell — APPROVED** [review]

**Review by weimiao67 — APPROVED** [review]
